### ISSUE TITLE
[FEATURE] Add support for the AWS OpenSearch Ingestion Pipeline

### ISSUE THREAD

**Issue by adilnaimi** [opensearch-project/opensearch-py#405]
### Is your feature request related to a problem?
No

### What solution would you like?
Native access to the newly released AWS feature, [OpenSearch Ingestion Pipeline](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/ingestion.html) 

### What alternatives have you considered?
Our current workaround uses requests-aws4auth but unnecessarily adds another layer of complexity to our codebase.

### Do you have any additional context?
No

**Comment by saimedhi** [comment]
@wbeckler, please take a look at this feature request.

**Comment by wbeckler** [comment]
@adilnaimi Would you explain more how you would use the opensearch client for accessing the ingestion service?

**Comment by wbeckler** [comment]
I'm not sure where that request would belong either. It sounds like you're looking for a data-prepper version of this: https://github.com/vklochan/python-logstash

Maybe you could fork it or start from scratch and make the first client for data-prepper? If so, leave a comment here so if anyone else wants to help out they'll see this and find you.

**Comment by Utkarsh-Aga** [comment]
@wbeckler @adilnaimi - Just want to re-confirm my understanding that we are looking for a method which we can use with `opensearch.py` package to ingest the data instead of using the following call -
```
awscurl --service osis --region us-east-1 \
    -X POST \
    -H "Content-Type: application/json" \
    -d '[{"time":"2014-08-11T11:40:13+00:00","remote_addr":"122.226.223.69","status":"404","request":"GET http://www.k2proxy.com//hello.html HTTP/1.1","http_user_agent":"Mozilla/4.0 (compatible; WOW64; SLCC2;)"}]' \
    https://{pipeline-endpoint}.us-east-1.osis.amazonaws.com/log-pipeline/test_ingestion_path
```

Just like using `client.cat.indices` instead of `curl https://{domain-endpoint}/_cat/indices` 

I ask this because if my understanding is correct, I would be interested to be part of this.

**Comment by dblock** [comment]
@Utkarsh-Aga do you think it should be some kind of option, new namespace, require a separate client instance, or be a separate client altogether because it's specific to an AWS service?

- `client = Client(ingestion_pipeline: "https://{pipeline-endpoint}.us-east-1.osis.amazonaws.com/log-pipeline/test_ingestion_path")`
- `client.ingestion_pipeline(...)...`
- something else?

**Comment by Utkarsh-Aga** [comment]
@dblock I believe, while creating the authentication for the client we need to provide the service as `osis` and then we can leverage options like `client.ingestion_pipeline(...)`

**Comment by dblock** [comment]
So we would treat osis as a plugin? My question is whether we're better off writing another library that depends on opensearch-py? Either way, since the service is not available in open source, it should be behind import `aws.ingestion_pipeline`. 

Want to contribute @Utkarsh-Aga?

**Comment by Utkarsh-Aga** [comment]
@dblock - 
Yeah, we can have it as a plugin, because it would only have one functionality to send the data over http/https to the Ingestion Pipeline.
Further, I would be happy to contribute to it.
But I did not get this statement `it should be behind import aws.ingestion_pipeline.`, can you please elaborate a bit more on this ?

**Comment by dblock** [comment]
> Yeah, we can have it as a plugin, because it would only have one functionality to send the data over http/https to the Ingestion Pipeline. Further, I would be happy to contribute to it.

Awesome.

> But I did not get this statement it should be behind import aws.ingestion_pipeline., can you please elaborate a bit more on this ?

Since it's an AWS service, and not a generic feature of OpenSearch open source, you can't treat it like other plugins. We can't have `client.ingestion_pipeline`, it would need to be `client.aws_ingestion_pipeline`, but that's really ugly. So as a developer I think I'd like to be able to do something like this:

```python
from opensearch.aws import IngestionPipeline

client = Client(
  plugins: IngestionPipeline
)
client.ingestion_pipeline....
```

Does it make sense?

**Comment by Utkarsh-Aga** [comment]
Got it, Thanks a lot for these details @dblock, it made things super clear.
I would start my research on this on how to implement it.

Since, I would be contributing to this first time, so I should just follow - [CONTRIBUTING](https://github.com/opensearch-project/opensearch-py/blob/main/CONTRIBUTING.md) guide?

**Comment by dblock** [comment]
Yes! Let us know if you need help. A good place to ask general questions is the public Slack - https://opensearch.org/slack.html

**Comment by Utkarsh-Aga** [comment]
Sure, Thanks.

**Comment by dlvenable** [comment]
@adilnaimi , @Utkarsh-Aga ,

As noted above, there is an existing `http` source you can use. But, it seems you are interested in having APIs that look similar to OpenSearch APIs. We have an existing issue to support the `_bulk` API - https://github.com/opensearch-project/data-prepper/issues/248. Another feature we've discussed is having an API that looks like existing OpenSearch document APIs. For example, `POST {index}/_doc`. Is this something along of the lines of what you are looking for?

**Comment by Utkarsh-Aga** [comment]
Hello @dlvenable - Yes, if we have a support for `_bulk` or `_doc`, then it might help, and we need not create a different plugin in `opensearch.py` to send the data to `http` source.
Would let @adilnaimi also once confirm, if that is the same thing, they were looking for?

**Comment by jzonthemtn** [comment]
This feature would be helpful to the OpenSearch [UBI](https://github.com/opensearch-project/OpenSearch/issues/12084) efforts. As [commented](https://github.com/opensearch-project/data-prepper/issues/248#issuecomment-2093859216) in Data Prepper, being able to utilize an OpenSearch-like API to index UBI events and queries through Data Prepper would provide the user flexibility for storing UBI data with low overhead.
